# Capstone: Find the best neighborhood in Toronto to open a Restaurant Supply Store

##  Load all the Data from all the various sources.

###  Toronto neighborhoods broken down by postal code.

In [1]:
# Load the required libraries
import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup

res = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
soup = BeautifulSoup(res.content,'lxml')
table = soup.find_all('table')[0] 
df = pd.read_html(str(table))

data = pd.DataFrame(df[0])
data = data.rename(columns={0:'PostalCode', 1:'Bourough', 2:'Neighborhood'})
data = data.iloc[1:]
data = data[~data['Bourough'].str.contains('Not assigned')]
df2=data.groupby(['PostalCode', 'Bourough']).apply(lambda group: ', '.join(group['Neighborhood']))
df2=df2.to_frame().reset_index()
df2 = df2.rename(columns={0:'Neighborhood'})
df2.loc[df2.Neighborhood == 'Not assigned', 'Neighborhood' ] = df2.Bourough
df2.head()   

PostalCode     Bourough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

###  Load the Toronto geospatial cooridinates and merge to Toronto Postal Code Data

In [2]:
!wget -O to_geo_space.csv http://cocl.us/Geospatial_data
gf = pd.read_csv('to_geo_space.csv')
gf = gf.rename(columns={'Postal Code':'PostalCode'})
gf_new = pd.merge(df2, gf, on='PostalCode', how='inner')
gf_new.head()


--2019-01-19 16:06:44--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 169.48.113.201
Connecting to cocl.us (cocl.us)|169.48.113.201|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2019-01-19 16:06:44--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|169.48.113.201|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2019-01-19 16:06:48--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.27.197
Connecting to ibm.box.com (ibm.box.com)|107.152.27.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.ent.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2019-01-19 16:06:48--  https://ibm.ent.box.com/shared/

PostalCode     Bourough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

###  Toronto neighborhoods populations broken down by postal code

In [3]:
df_pop = pd.read_csv('https://www12.statcan.gc.ca/census-recensement/2016/dp-pd/hlt-fst/pd-pl/Tables/File.cfm?T=1201&SR=1&RPP=9999&PR=0&CMA=0&CSD=0&S=22&O=A&Lang=Eng&OFT=CSV',encoding = 'unicode_escape')
df_pop = df_pop.rename(columns={'Geographic code':'PostalCode', 'Geographic name':'PostalCod2', 'Province or territory':'Province', 'Incompletely enumerated Indian reserves and Indian settlements, 2016':'Incomplete', 'Population, 2016':'Population_2016', 'Total private dwellings, 2016':'TotalPrivDwellings', 'Private dwellings occupied by usual residents, 2016':'PrivDwellingsOccupied'})
df_pop= df_pop.drop(columns=['PostalCod2', 'Province', 'Incomplete', 'TotalPrivDwellings', 'PrivDwellingsOccupied'])
df_pop = df_pop.iloc[1:]
df_pop.head()

PostalCode  Population_2016
1        A0A          46587.0
2        A0B          19792.0
3        A0C          12587.0
4        A0E          22294.0
5        A0G          35266.0

###  Merge Toronto Neighbourhood populations data with Toronto Postal Code data

In [4]:
gf_new
gf_new = pd.merge(df_pop, gf_new, on='PostalCode', how='right')
ues(by=['Population_2016'], ascending=False)
gf_new.head()

PostalCode  Population_2016     Bourough  \
22         M2N          75897.0   North York   
0          M1B          66108.0  Scarborough   
18         M2J          58293.0   North York   
100        M9V          55959.0    Etobicoke   
14         M1V          54680.0  Scarborough   

                                          Neighborhood   Latitude  Longitude  
22                                    Willowdale South  43.770120 -79.408493  
0                                       Rouge, Malvern  43.806686 -79.194353  
18                        Fairview, Henry Farm, Oriole  43.778517 -79.346556  
100  Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437  
14   Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577

*** Key Observation here is a list of Toronto Neighborhoods Populations  ***

###  Toronto neighborhoods average after tax income broken down by postal code

In [5]:
df_income = pd.read_csv('TorontoAvgIncomeByPC.csv',encoding = 'unicode_escape')
df_income = df_income.rename(columns={"after-tax income of households in 2015":"AfterTaxIncome2015"})
df_income.head()

PostalCode AfterTaxIncome2015
0        M1A               Null
1        M2A               Null
2        M3A              57698
3        M4A              49453
4        M5A              46938

###  Merge Toronto Neighbourhood income data with Toronto Postal Code data

In [6]:
gf_new = pd.merge(df_income, gf_new, on='PostalCode', how='right')
gf_new = gf_new.replace('Null', 0)
gf_new['AfterTaxIncome2015'] = gf_new['AfterTaxIncome2015'].astype('float64') 
gf_new = gf_new.sort_values(by=['AfterTaxIncome2015'], ascending=False)
gf_new.to_csv('TO_Affluence.csv')
gf_new.head(10)


PostalCode  AfterTaxIncome2015  Population_2016         Bourough  \
66        M2P            115237.0           7843.0       North York   
55        M5M            111821.0          25975.0       North York   
61        M4N            109841.0          15330.0  Central Toronto   
74        M5R            108271.0          26496.0  Central Toronto   
98        M8X             97210.0          10787.0        Etobicoke   
45        M2L             96512.0          11717.0       North York   
23        M4G             94853.0          19076.0        East York   
12        M1C             93943.0          35626.0      Scarborough   
11        M9B             91110.0          32400.0        Etobicoke   
7         M3B             90841.0          13324.0       North York   

                                         Neighborhood   Latitude  Longitude  
66                                    York Mills West  43.752758 -79.400049  
55                  Bedford Park, Lawrence Manor East  43.733283 -79.419750  
61                                      Lawrence Park  43.728020 -79.388790  
74                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
98      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
45                           Silver Hills, York Mills  43.757490 -79.374714  
23                                            Leaside  43.709060 -79.363452  
12             Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
11  Cloverdale, Islington, Martin Grove, Princess ...  43.650943 -79.554724  
7                                     Don Mills North  43.745906 -79.352188

###  What is the Canadian National Average After Tax Income.

Again obtained from the Stats Canada Website
Canadian families and unattached individuals had a median after-tax income of $57,000 in 2016. 

###  Toronto list of Restaurants or Venues that could potentially use Restaurant Equipment.

In [7]:
#FourSquare Credentials

CLIENT_ID = 'NUMEDNFTTRUEWOQRQL44JMMFU1TQWT1YXVXSLVQUKRYHCF0B' # your Foursquare ID


CLIENT_SECRET = '0JK0YSMUGX2SUL0DAGRSFDAN5JS1PUTPZIRTTDZWTGJ15SR2' # your Foursquare Secret


VERSION = '20180604' # Foursquare API version

In [8]:

import requests 
from pandas.io.json import json_normalize 

LIMIT = 200 

radius = 500

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
       
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
       
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [9]:
# Toronto Bouroughs
TO_data = gf_new
TO_data.head()

PostalCode  AfterTaxIncome2015  Population_2016         Bourough  \
66        M2P            115237.0           7843.0       North York   
55        M5M            111821.0          25975.0       North York   
61        M4N            109841.0          15330.0  Central Toronto   
74        M5R            108271.0          26496.0  Central Toronto   
98        M8X             97210.0          10787.0        Etobicoke   

                                     Neighborhood   Latitude  Longitude  
66                                York Mills West  43.752758 -79.400049  
55              Bedford Park, Lawrence Manor East  43.733283 -79.419750  
61                                  Lawrence Park  43.728020 -79.388790  
74            The Annex, North Midtown, Yorkville  43.672710 -79.405678  
98  The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944

###  Get all the Venues in Toronto.

In [11]:
# Get all of the Venues
TO_venues = getNearbyVenues(names=TO_data['Neighborhood'],
                                   latitudes=TO_data['Latitude'],
                                   longitudes=TO_data['Longitude']
                                  )

York Mills West
Bedford Park, Lawrence Manor East
Lawrence Park
The Annex, North Midtown, Yorkville
The Kingsway, Montgomery Road, Old Mill North
Silver Hills, York Mills
Leaside
Highland Creek, Rouge Hill, Port Union
Cloverdale, Islington, Martin Grove, Princess Gardens, West Deane Park
Don Mills North
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
East Toronto
Upper Rouge
Birch Cliff, Cliffside West
Little Portugal, Trinity
Rosedale
Moore Park, Summerhill East
Roselawn
The Beaches
Kingsway Park South West, Mimico NW, The Queensway West, Royal York South West, South of Bloor
North Toronto West
Runnymede, Swansea
Berczy Park
Forest Hill North, Forest Hill West
Woodbine Heights
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Harbourfront East, Toronto Islands, Union Station
Adelaide, King, Richmond
St. James Town
The Beaches West, India Bazaar
Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe
Kings

Let's count the number of Venues per Neighorhood

In [12]:
TO_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     10   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        17   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            54   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      6   
Brockton, Exhibition Place, Parkdale Village                           19   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto, Downsview East                                             4   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            47   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               8   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                 9   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Cloverdale, Islington, Martin Grove, Princess G...                      2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                       10   
...                                                                   ...   
Northwest                                                               2   
Northwood Park, York University                                         5   
Parkdale, Roncesvalles                                                 16   
Parkwoods                                                               3   
Queen's Park                                                           44   
Rosedale                                                                4   
Roselawn                                                                3   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     40   
Ryerson, Garden District                                              100   
Scarborough Village                                                     2   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         20   
The Danforth West, Riverdale                                           42   
The Junction North, Runnymede                                           4   
The Kingsway, Montgomery Road, Old Mill North   

In [13]:
# Let's pick out restaurants from Venue Categories

print('Unique Venue Categories:')
list(TO_venues['Venue Category'].unique())

Unique Venue Categories:


['Bank',
 'Park',
 'Electronics Store',
 'Café',
 'Hardware Store',
 'Sushi Restaurant',
 'Indian Restaurant',
 'Restaurant',
 'Coffee Shop',
 'Pub',
 'Italian Restaurant',
 'Liquor Store',
 'Comfort Food Restaurant',
 'Juice Bar',
 'Grocery Store',
 'Butcher',
 'Greek Restaurant',
 'Thai Restaurant',
 'Sandwich Place',
 'American Restaurant',
 'Pharmacy',
 'Fast Food Restaurant',
 'Pizza Place',
 'Dim Sum Restaurant',
 'Swim School',
 'Bus Line',
 'Vegetarian / Vegan Restaurant',
 'Burger Joint',
 'BBQ Joint',
 'History Museum',
 'Jewish Restaurant',
 'Cosmetics Shop',
 'Furniture / Home Store',
 'Pool',
 'River',
 'Sports Bar',
 'Fish & Chips Shop',
 'Sporting Goods Shop',
 'Gym',
 'Bike Shop',
 'Supermarket',
 'Pet Store',
 'Smoothie Shop',
 'Shopping Mall',
 'Clothing Store',
 'Dessert Shop',
 'Breakfast Spot',
 'Brewery',
 'Mexican Restaurant',
 'Beer Store',
 'Bagel Shop',
 'Convenience Store',
 'Bar',
 'Moving Target',
 'Golf Course',
 'Gym / Fitness Center',
 'Caribbean Restaur

###  Only add Restaurants as Venue Categories

In [14]:
# Here we manually pick out restaurants or 'features' from the unique venue list and that we want to examine for similiarity during clustering
rest_list = ['Steakhouse', 'Coffee Shop', 'Café', 'Ramen Restaurant', 'Indonesian Restaurant', 'Restaurant', 'Japanese Restaurant', 
             'Fast Food Restaurant', 'Sushi Restaurant', 'Vietnamese Restaurant', 'Pizza Place', 'Sandwich Place', 'Middle Eastern Restaurant', 
             'Burger Joint', 'American Restaurant', 'Food Court', 'Wings Joint', 'Burrito Place', 'Asian Restaurant', 'Deli / Bodega', 
             'Greek Restaurant', 'Fried Chicken Joint', 'Airport Food Court', 'Chinese Restaurant', 'Breakfast Spot', 'Mexican Restaurant',
             'Indian Restaurant', 'Latin American Restaurant', 'Bar', 'Pub', 'Italian Restaurant', 'French Restaurant', 'Ice Cream Shop', 
             'Caribbean Restaurant', 'Gastropub', 'Thai Restaurant', 'Cajun / Creole Restaurant', 'Diner', 'Dim Sum Restaurant', 'Seafood Restaurant', 
             'Food & Drink Shop', 'Noodle House', 'Food', 'Fish & Chips Shop', 'Falafel Restaurant', 'Gourmet Shop', 'Vegetarian / Vegan Restaurant', 
             'South American Restaurant', 'Korean Restaurant', 'Cuban Restaurant', 'New American Restaurant', 'Malay Restaurant', 'Mac & Cheese Joint',
             'Bistro', 'Southern / Soul Food Restaurant', 'Tapas Restaurant',  'Sports Bar', 'Polish Restaurant', 'Ethiopian Restaurant', 
             'Creperie', 'Sake Bar', 'Persian Restaurant', 'Afghan Restaurant','Mediterranean Restaurant', 'BBQ Joint', 'Jewish Restaurant', 
             'Comfort Food Restaurant',  'Hakka Restaurant', 'Food Truck', 'Taiwanese Restaurant',  'Snack Place', 'Eastern European Restaurant', 
             'Dumpling Restaurant', 'Belgian Restaurant', 'Arepa Restaurant', 'Taco Place', 'Doner Restaurant', 'Filipino Restaurant', 
             'Hotpot Restaurant', 'Poutine Place', 'Salad Place',  'Portuguese Restaurant', 'Modern European Restaurant', 'Empanada Restaurant', 
             'Irish Pub', 'Molecular Gastronomy Restaurant', 'German Restaurant', 'Brazilian Restaurant', 'Gluten-free Restaurant', 'Soup Place']

rest_pd = pd.DataFrame(rest_list)

rest_pd = rest_pd.rename(columns={0:'Venue Category'})


TO_new = pd.merge(TO_venues, rest_pd, on='Venue Category', how='right')


TO_new.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                               60   
Agincourt                                                               2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      5   
Alderwood, Long Branch                                                  5   
Bathurst Manor, Downsview North, Wilson Heights                        10   
Bayview Village                                                         3   
Bedford Park, Lawrence Manor East                                      18   
Berczy Park                                                            26   
Birch Cliff, Cliffside West                                             1   
Bloordale Gardens, Eringate, Markland Wood, Old...                      2   
Brockton, Exhibition Place, Parkdale Village                           10   
Business Reply Mail Processing Centre 969 Eastern                       4   
CN Tower, Bathurst Quay, Island airport, Harbou...                      1   
Cabbagetown, St. James Town                                            28   
Caledonia-Fairbanks                                                     1   
Canada Post Gateway Processing Centre                                   8   
Cedarbrae                                                               4   
Central Bay Street                                                     60   
Chinatown, Grange Park, Kensington Market                              65   
Christie                                                                7   
Church and Wellesley                                                   55   
Clairlea, Golden Mile, Oakridge                                         1   
Clarks Corners, Sullivan, Tam O'Shanter                                 8   
Cliffcrest, Cliffside, Scarborough Village West                         1   
Commerce Court, Victoria Hotel                                         67   
Davisville                                                             25   
Davisville North                                                        4   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     10   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          3   
Design Exchange, Toronto Dominion Centre                               68   
...                                                                   ...   
Kingsway Park South West, Mimico NW, The Queens...                      4   
L'Amoreaux West, Steeles West                                           9   
Lawrence Heights, Lawrence Manor                                        2   
Lawrence Park                                                           1   
Leaside                                                                12   
Little Portugal, Trinity                                               42   
Maryvale, Wexford                                                       2   
North Toronto West                                                      7   
Northwood Park, York University                                         2   
Parkdale, Roncesvalles                                                  9   
Parkwoods                                                               2   
Queen's Park                                                           31   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     30   
Ryerson, Garden District                                               50   
St. James Town                                                         53   
Stn A PO Boxes 25 The Esplanade                                        52   
Studio District                                                        22   
The Annex, North Midtown, Yorkville                                    18   
The Beaches                                     

###  OneHot encode and count restaurants

In [15]:

TO_new_onehot = pd.get_dummies(TO_new[['Venue Category']], prefix="", prefix_sep="")
TO_new_onehot['Neighborhood'] = TO_new['Neighborhood'] 
fixed_columns = [TO_new_onehot.columns[-1]] + list(TO_new_onehot.columns[:-1])
TO_new_onehot = TO_new_onehot[fixed_columns]

TO_new_onehot.head()


Neighborhood  Afghan Restaurant  Airport Food Court  \
0    Bedford Park, Lawrence Manor East                  0                   0   
1  The Annex, North Midtown, Yorkville                  0                   0   
2  The Annex, North Midtown, Yorkville                  0                   0   
3  The Annex, North Midtown, Yorkville                  0                   0   
4                      Don Mills North                  0                   0   

   American Restaurant  Arepa Restaurant  Asian Restaurant  BBQ Joint  Bar  \
0                    0                 0                 0          0    0   
1                    0                 0                 0          0    0   
2                    0                 0                 0          0    0   
3                    0                 0                 0          0    0   
4                    0                 0                 0          0    0   

   Belgian Restaurant  Bistro     ...       Sports Bar  Steakhouse  \
0                   0       0     ...                0           0   
1                   0       0     ...                0           0   
2                   0       0     ...                0           0   
3                   0       0     ...                0           0   
4                   0       0     ...                0           0   

   Sushi Restaurant  Taco Place  Taiwanese Restaurant  Tapas Restaurant  \
0                 0           0                     0                 0   
1                 0           0                     0                 0   
2                 0           0                     0                 0   
3                 0           0                     0                 0   
4                 0           0                     0                 0   

   Thai Restaurant  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  \
0                0                              0                      0   
1                0                              0                      0   
2                0                              0                      0   
3                0                              0                      0   
4                0                              0                      0   

   Wings Joint  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 91 columns]

In [16]:
#Analyze each neighbourhood



TO_grouped = TO_new_onehot.groupby('Neighborhood').mean().reset_index()
TO_grouped.shape


TO_grouped.head()




Neighborhood  Afghan Restaurant  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
3                             Alderwood, Long Branch                0.0   
4    Bathurst Manor, Downsview North, Wilson Heights                0.0   

   Airport Food Court  American Restaurant  Arepa Restaurant  \
0                 0.0             0.066667               0.0   
1                 0.0             0.000000               0.0   
2                 0.0             0.000000               0.0   
3                 0.0             0.000000               0.0   
4                 0.0             0.000000               0.0   

   Asian Restaurant  BBQ Joint   Bar  Belgian Restaurant  Bistro     ...       \
0              0.05        0.0  0.05                 0.0     0.0     ...        
1              0.00        0.0  0.00                 0.0     0.0     ...        
2              0.00        0.0  0.00                 0.0     0.0     ...        
3              0.00        0.0  0.00                 0.0     0.0     ...        
4              0.00        0.0  0.00                 0.0     0.0     ...        

   Sports Bar  Steakhouse  Sushi Restaurant  Taco Place  Taiwanese Restaurant  \
0         0.0    0.066667          0.033333         0.0                   0.0   
1         0.0    0.000000          0.000000         0.0                   0.0   
2         0.0    0.000000          0.000000         0.0                   0.0   
3         0.0    0.000000          0.000000         0.0                   0.0   
4         0.0    0.000000          0.100000         0.0                   0.0   

   Tapas Restaurant  Thai Restaurant  Vegetarian / Vegan Restaurant  \
0               0.0         0.066667                       0.016667   
1               0.0         0.000000                       0.000000   
2               0.0         0.000000                       0.000000   
3               0.0         0.000000                       0.000000   
4               0.0         0.000000                       0.000000   

   Vietnamese Restaurant  Wings Joint  
0                    0.0          0.0  
1                    0.0          0.0  
2                    0.0          0.0  
3                    0.0          0.0  
4                    0.0          0.0  

[5 rows x 91 columns]

## Start Clustering

### Use silhouette score to find optimal number of clusters to segment the data

In [17]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import numpy as np


TO_grouped_clustering = TO_grouped.drop('Neighborhood', 1)


kclusters = np.arange(2,10)
results = {}
for size in kclusters:
    model = KMeans(n_clusters = size).fit(TO_grouped_clustering)
    predictions = model.predict(TO_grouped_clustering)
    results[size] = silhouette_score(TO_grouped_clustering, predictions)

best_size = max(results, key=results.get)
best_size

2

### Run K means and segment data into clusters and generate labels

In [18]:

from sklearn.cluster import KMeans

kclusters = best_size

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(TO_grouped_clustering)

kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [19]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']


columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = TO_grouped['Neighborhood']

for ind in np.arange(TO_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(TO_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Albion Gardens, Beaumond Heights, Humbergate, ...  Fast Food Restaurant   
3                             Alderwood, Long Branch           Pizza Place   
4    Bathurst Manor, Downsview North, Wilson Heights           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1        Sandwich Place     Fish & Chips Shop      Doner Restaurant   
2   Fried Chicken Joint           Coffee Shop        Sandwich Place   
3                   Pub           Coffee Shop        Sandwich Place   
4  Fast Food Restaurant        Sandwich Place           Pizza Place   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                   Restaurant      Asian Restaurant   
1   Dumpling Restaurant  Eastern European Restaurant   Empanada Restaurant   
2           Pizza Place                  Wings Joint                 Diner   
3    Falafel Restaurant           Dim Sum Restaurant                 Diner   
4                 Diner          Fried Chicken Joint      Sushi Restaurant   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0                   Bar      Sushi Restaurant                  Pizza Place  
1  Ethiopian Restaurant    Falafel Restaurant         Fast Food Restaurant  
2      Doner Restaurant   Dumpling Restaurant  Eastern European Restaurant  
3      Doner Restaurant   Dumpling Restaurant  Eastern European Restaurant  
4            Restaurant         Deli / Bodega             Doner Restaurant

###  Merge the Toronto data with geo coordinates data and make sure it's the right shape

In [20]:

TO_labels = pd.merge(TO_data,TO_grouped, on='Neighborhood', how='right')
TO_labels.shape


TO_labels = TO_labels.drop(columns=['Steakhouse', 'Coffee Shop', 'Café', 'Ramen Restaurant', 'Indonesian Restaurant', 'Restaurant', 'Japanese Restaurant', 
             'Fast Food Restaurant', 'Sushi Restaurant', 'Vietnamese Restaurant', 'Pizza Place', 'Sandwich Place', 'Middle Eastern Restaurant', 
             'Burger Joint', 'American Restaurant', 'Food Court', 'Wings Joint', 'Burrito Place', 'Asian Restaurant', 'Deli / Bodega', 
             'Greek Restaurant', 'Fried Chicken Joint', 'Airport Food Court', 'Chinese Restaurant', 'Breakfast Spot', 'Mexican Restaurant',
             'Indian Restaurant', 'Latin American Restaurant', 'Bar', 'Pub', 'Italian Restaurant', 'French Restaurant', 'Ice Cream Shop', 
             'Caribbean Restaurant', 'Gastropub', 'Thai Restaurant', 'Cajun / Creole Restaurant', 'Diner', 'Dim Sum Restaurant', 'Seafood Restaurant', 
             'Food & Drink Shop', 'Noodle House', 'Food', 'Fish & Chips Shop', 'Falafel Restaurant', 'Gourmet Shop', 'Vegetarian / Vegan Restaurant', 
             'South American Restaurant', 'Korean Restaurant', 'Cuban Restaurant', 'New American Restaurant', 'Malay Restaurant', 'Mac & Cheese Joint',
             'Bistro', 'Southern / Soul Food Restaurant', 'Tapas Restaurant',  'Sports Bar', 'Polish Restaurant', 'Ethiopian Restaurant', 
             'Creperie', 'Sake Bar', 'Persian Restaurant', 'Afghan Restaurant','Mediterranean Restaurant', 'BBQ Joint', 'Jewish Restaurant', 
             'Comfort Food Restaurant',  'Hakka Restaurant', 'Food Truck', 'Taiwanese Restaurant',  'Snack Place', 'Eastern European Restaurant', 
             'Dumpling Restaurant', 'Belgian Restaurant', 'Arepa Restaurant', 'Taco Place', 'Doner Restaurant', 'Filipino Restaurant', 
             'Hotpot Restaurant', 'Poutine Place', 'Salad Place',  'Portuguese Restaurant', 'Modern European Restaurant', 'Empanada Restaurant', 
             'Irish Pub', 'Molecular Gastronomy Restaurant', 'German Restaurant', 'Brazilian Restaurant', 'Gluten-free Restaurant', 'Soup Place'])
TO_labels.head()

PostalCode  AfterTaxIncome2015  Population_2016         Bourough  \
0        M5M            111821.0          25975.0       North York   
1        M4N            109841.0          15330.0  Central Toronto   
2        M5R            108271.0          26496.0  Central Toronto   
3        M4G             94853.0          19076.0        East York   
4        M1C             93943.0          35626.0      Scarborough   

                             Neighborhood   Latitude  Longitude  
0       Bedford Park, Lawrence Manor East  43.733283 -79.419750  
1                           Lawrence Park  43.728020 -79.388790  
2     The Annex, North Midtown, Yorkville  43.672710 -79.405678  
3                                 Leaside  43.709060 -79.363452  
4  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497

###  Add the KMeans Labels

In [21]:
TO_merged = TO_labels

TO_merged['Cluster Labels'] = kmeans.labels_

TO_merged = TO_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

TO_merged.head() 

PostalCode  AfterTaxIncome2015  Population_2016         Bourough  \
0        M5M            111821.0          25975.0       North York   
1        M4N            109841.0          15330.0  Central Toronto   
2        M5R            108271.0          26496.0  Central Toronto   
3        M4G             94853.0          19076.0        East York   
4        M1C             93943.0          35626.0      Scarborough   

                             Neighborhood   Latitude  Longitude  \
0       Bedford Park, Lawrence Manor East  43.733283 -79.419750   
1                           Lawrence Park  43.728020 -79.388790   
2     The Annex, North Midtown, Yorkville  43.672710 -79.405678   
3                                 Leaside  43.709060 -79.363452   
4  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1  Fast Food Restaurant    Italian Restaurant   
1               1    Dim Sum Restaurant           Wings Joint   
2               1                  Café        Sandwich Place   
3               1           Coffee Shop          Burger Joint   
4               1                   Bar           Wings Joint   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0           Coffee Shop      Sushi Restaurant                  Pizza Place   
1      Doner Restaurant   Dumpling Restaurant  Eastern European Restaurant   
2           Coffee Shop           Pizza Place                    BBQ Joint   
3        Breakfast Spot    Mexican Restaurant             Sushi Restaurant   
4                  Food      Doner Restaurant          Dumpling Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                          Pub                  Café        Sandwich Place   
1          Empanada Restaurant  Ethiopian Restaurant    Falafel Restaurant   
2                 Burger Joint                   Pub     Indian Restaurant   
3                   Sports Bar     Fish & Chips Shop            Restaurant   
4  Eastern European Restaurant   Empanada Restaurant  Ethiopian Restaurant   

  9th Most Common Venue         10th Most Common Venue  
0     Indian Restaurant                     Restaurant  
1  Fast Food Restaurant            Filipino Restaurant  
2   American Restaurant  Vegetarian / Vegan Restaurant  
3        Sandwich Place           Fast Food Restaurant  
4    Falafel Restaurant           Fast Food Restaurant

In [22]:
TO_merged_new1 = TO_merged.loc[TO_merged['Cluster Labels'] == 0, TO_merged.columns[[3, 4] + list(range(5, TO_merged.shape[1]))]]
TO_merged_new1.shape

(4, 15)

In [23]:
TO_merged_new2 = TO_merged.loc[TO_merged['Cluster Labels'] == 1, TO_merged.columns[[3, 4] + list(range(5, TO_merged.shape[1]))]]
TO_merged_new2.shape

(76, 15)

##  Cluster 2 Contains the highest cluster density. We need to find the geographic centroid for this cluster. This is the optimum location for a new Restaurant Supply Store.

In [24]:

Cluster_0_coorid = TO_merged_new2[['Latitude', 'Longitude']]
Cluster_0_coorid = list(Cluster_0_coorid.values) 
lat = []
long = []



for l in Cluster_0_coorid:
  lat.append(l[0])
  long.append(l[1])



Blatitude = sum(lat)/len(lat)
Blongitude = sum(long)/len(long)
print(Blatitude)
print(Blongitude)

43.69907216578946
-79.38788743289474


###  Install opencage to reverse lookup the coordinates

In [25]:

!pip install opencage
from opencage.geocoder import OpenCageGeocode
from pprint import pprint

key = '1d97b344df184b1cb0d2427663f85ac6'
geocoder = OpenCageGeocode(key)

results = geocoder.reverse_geocode(Blatitude, Blongitude)
pprint(results)

[{'annotations': {'DMS': {'lat': "43° 41' 56.97528'' N",
                          'lng': "79° 23' 16.39356'' W"},
                  'MGRS': '17TPJ2990139722',
                  'Maidenhead': 'FN03hq37kt',
                  'Mercator': {'x': -8837419.167, 'y': 5389473.786},
                  'OSM': {'edit_url': 'https://www.openstreetmap.org/edit?way=401259844#map=17/43.69916/-79.38789',
                          'url': 'https://www.openstreetmap.org/?mlat=43.69916&mlon=-79.38789#map=17/43.69916/-79.38789'},
                  'callingcode': 1,
                  'currency': {'alternate_symbols': ['C$', 'CAD$'],
                               'decimal_mark': '.',
                               'disambiguate_symbol': 'C$',
                               'html_entity': '$',
                               'iso_code': 'CAD',
                               'iso_numeric': 124,
                               'name': 'Canadian Dollar',
                               'smallest_denomination': 5,
 

In [26]:
#Obtain the popupstring of the best location
popstring = TO_data[TO_data['PostalCode'].str.contains('M4S')]

def str_join(*args):
    return ''.join(map(str, args))

popstring_new = str_join('The Best Neighbourhood to locate a Restaurant Supply Store is in: ', popstring['Neighborhood'].values,  ' in ' ,  popstring['Bourough'].values)


print(popstring_new)


The Best Neighbourhood to locate a Restaurant Supply Store is in: ['Davisville'] in ['Central Toronto']


In [27]:
# Let's get the coordinates for Toronto

from geopy.geocoders import Nominatim
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


##  Results

###  Plot the clusters on a Map of the Toronto and Super Impose the best location of a Store

In [28]:

import folium 

import matplotlib.cm as cm
import matplotlib.colors as colors


map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)


x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


markers_colors = []
for lat, lon, poi, cluster in zip(TO_merged['Latitude'], TO_merged['Longitude'], TO_merged['Neighborhood'], TO_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
folium.CircleMarker([Blatitude, Blongitude],
                    radius=50,
                    popup='Toronto',
                    color='red',
                    ).add_to(map_clusters)


map_clusters.add_child(folium.ClickForMarker(popup=popstring_new))
       

map_clusters.save('map_clusters.html')

![](https://i.imgur.com/lQnSq7P.png)

###  Exact Address of desired Location

In [29]:
print('The exact Address to locate would be: 268 Balliol Street, ON M4S 1C2, Canada or lat: 43.6991598, lng: -79.3878871')

The exact Address to locate would be: 268 Balliol Street, ON M4S 1C2, Canada or lat: 43.6991598, lng: -79.3878871


## Discussion:

### Explaining the results

As we built our list of neighborhoods with Restaurant venues exclusively we discovered most neighborhoods were similar and the greatest concentration of restaurants was in Central Toronto and downtown Toronto. This might seem obvious but it would also appear that these are some of the most affluent neighborhoods in Toronto so there appears to be correlation. By Locating in the general vicinity of the Exact location my friend could be geographically centered in this cluster and poised to service his restaurant customer base with the greatest efficiency.<br>

When we built our our K-Means dataset we used Silhouette analysis to tell us there was a lot of similarity between neighborhoods and the most common restaurants contained with in. Really there was only 2 types of cluster or neighborhoods in greater Toronto. The vast majority of those were in 1 cluster. So Toronto restaurants might be many but they are very homogeneously located near the center of Toronto.<br>

Of the 103 Toronto Neighborhoods gathered only 55.3% or 57 Neighborhoods are above the median after-tax income. 37.8% or 39 Neighborhoods are below he median after-tax income. 6.7% or 7 neighborhoods did not register as it appears their populations are too low. It appears that the greatest concentration of affluence is near central Toronto. We decided to keep all neighborhoods in the dataset regardless of income of population as the majority were close enough.<br> 




## Conclusion:

I feel confident with the recommendation I have given my friend as it is backed up with demonstrated data analysis. While nothing can ever be 100% certain he will certainly be better informed than he was prior to asking for my help.

Much more inference can be obtained with more work. A potential side business for my friend might be assisting new restaurant owners where they might locate a new restaurant, who their competition is and who their clientele might be.